docs/JLBP-0021.md

```diff
@@ -0,0 +1,28 @@
+---
+jlbp:
+  id: JLBP-21
+permalink: /JLBP-21
+---
+# Upload artifacts to Maven Central
```

Comment:
Did you have specific problems in your mind when you wrote this? I want to see it in this document.

Reply:
Typically it's happened with things like Gradle and SBT. I'd have to spend some time digging up specific cases.

Reply:
Looking forward.

Reply:
one example: https://github.com/grpc/grpc-java/issues/7276

Reply:
Thanks. If I read that correctly grpc-java have always published the library to Maven Central, but not always with JCenter. So this problem cannot be fixed by uploading some artifacts to Maven Central.

Reply:
I don't think that's true. Regardless, we've hit this repeatedly, every few months, in one context or another. I don't think we should list every example here, which might be resolved in the future.

Reply:
(You definitely have troubleshooted more cases than I have.) No need to list every occurrence. I want to learn just one case where this JLBP "Upload artifacts to Maven Central" would have prevented it.

Reply:
Ktor is a Kotlin library. https://ktor.io/docs/quickstart-index.html#set-up-a-ktor-project I guess jcenter is Kotlin's standard.

Reply:
That Ktor involves Kotlin is not relevant. Someone tried to use this library from Maven and it failed because it wasn't in Maven Central. This is a very common occurrence. I've encountered it multiple times myself. There are many different causes for it, but one solution. Deploy to Maven Central.

Reply:
Yes, one solution is for the Kotlin library maintainers to deploy the library to Maven central. Another solution, which I believe most common, is to include repositories in pom.xml.

From Java developers' perspective, having everything in Maven Central is simpler. I just feel that it's strange to recommend Kotlin library maintainers to publish their libraries to Maven Central when their standard is to use jcenter.

Reply:
It does not seem to be the Kotlin standard to use JCenter. Regardless, I have no compunctions about recommending they use Maven Central instead. There are many, many cases where using other repos causes needless pain for developers. None of these are unfixable, but there's no reason to make developers worry about this when there's such a simple solution. I do not want to litigate each and every instance of this problem.

Reply:
Are you aware of: https://maven.google.com/web/index.html ??

Reply:
Yes. I think we've had problems with android dependencies in the past for the reasons addressed in this JLBP. Tools start to break when they don't find artifacts in Maven Central. Yes there's usually some way to fix it, but the JLBPs don't stop at what works if you toggle the right knobs and flip the right switches. The JLBPs recommend practices that make development as easy as possible for clients. It's easier if the artifacts are in maven Central, the default where everything looks, than if they're in some other repo.

Reply:
And still another case of this that bit me today in one of our own repos:

```
[ERROR] Failed to execute goal on project google-api-client-android: Could not resolve dependencies for project com.google.api-client:google-api-client-android:jar:1.31.2-SNAPSHOT: Could not find artifact com.google.android.google-play-services:google-play-services:jar:1 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
```

I stand behind this one and I think we should publish it. Not uploading jars to Maven Central can be made to work, but that's not enough. Everyone's job is easier if libraries put their jars in Maven Central.

Reply:
One part of Google (this PR JLBP 21) is going to say everyone should put artifacts to Maven Central. Another part of Google may continue to use https://maven.google.com/web/index.html.

Do you think it's worth getting input from Android team?

Reply:
Feel free to ask them to comment on this PR. However in general we should not assume that just because Google does something, it is ideal. Teams often optimize for local convenience and development velocity, even when it costs external developers more time and effort. The JLBPs are about how to publish artifacts in a way that is easiest for customers to consume, even at the cost of extra effort for library developers.